test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,179 @@
+import os
+import shutil
+import tempfile
+from string import Template
+
+import torch
+import common_utils as common
+from common_nn import module_tests
+import torch.utils.cpp_extension
+from cpp_api_parity import sample_module
+
+torch_nn_has_parity = ['Linear']
+
+
+class TestCppApiParity(common.TestCase):
+    def setUp(self):
+        default_build_root = torch.utils.cpp_extension.get_default_build_root()
+        if os.path.exists(default_build_root):
+            shutil.rmtree(default_build_root)
+
+    def _test_torch_nn_modules(
+            self,
+            test_suite_name,
+            module_names,
+            module_tests,
+            python_module_class=None,
+            cpp_namespace='torch::nn::'):
+        torch_nn_test_methods = ['init', 'forward']
+
+        TORCH_NN_MODULE_COMMON_TEST_HARNESS = """\n
+const char * const parity_test_error_msg = "Parity test failed";
+
+void test_module_state_equality(std::shared_ptr<torch::nn::Module> m1, std::shared_ptr<torch::nn::Module> m2) {
+  auto params1 = m1->named_parameters();
+  auto params2 = m2->named_parameters();
+  TORCH_CHECK(
+    params1.size() == params2.size(),
+    parity_test_error_msg, ": ", "# of parameters doesn't match");
+  for (auto& param : params1)
+    TORCH_CHECK(
+      param.value().allclose(params2[param.key()]),
+      parity_test_error_msg, ": ", "value of `", param.key(), "` doesn't match");
+
+  auto buffers1 = m1->named_buffers();
+  auto buffers2 = m2->named_buffers();
+  TORCH_CHECK(
+    buffers1.size() == buffers2.size(),
+    parity_test_error_msg, ": ", "# of buffers doesn't match");
+  for (auto& buffer : buffers1)
+    TORCH_CHECK(
+      buffer.value().allclose(buffers2[buffer.key()]),
```

Comment:
`allclose` broadcasts. It would probably be good to check that the sizes of everything is the same.

Reply:
Fixed